File: rust/protocols/src/pkt_buffer.rs
#![allow(non_camel_case_types)]

pub struct netos_pkt_buffer {
    pub buffer : [u8; 4096 * 16],
    pub offset : usize,
    pub rx_len : usize,
}

impl netos_pkt_buffer {
    pub fn new() -> Self {
        Self {
            buffer : [0; 4096 * 16],
            offset : 0,
            rx_len : 0,
        }
    }

    pub fn has_short_len(&mut self) -> bool {
        self.offset >= self.rx_len
    }

    pub fn decode_2_bytes(&mut self) -> u16 {
        let u16_val = ((self.buffer[self.offset] as u16) << 8) |
                       (self.buffer[self.offset + 1]) as u16;
        self.offset += 2;

        u16_val
    }

    pub fn encode_2_bytes(&mut self, val : u16) {
        self.buffer[self.offset]        = ((val & 0xFF00) >> 8) as u8;
        self.buffer[self.offset + 1]    = (val & 0x00FF) as u8;
    }

    pub fn encode_4_bytes(&mut self, val : u32) {
        self.buffer[self.offset]        = ((val & 0xFF000000) >> 24) as u8;
        self.buffer[self.offset + 1]    = ((val & 0x00FF0000) >> 16) as u8;
        self.buffer[self.offset + 2]    = ((val & 0x0000FF00) >> 8) as u8;
        self.buffer[self.offset + 3]    = (val & 0x000000FF) as u8;
    }

    pub fn decode_4_bytes(&mut self) -> u32 {
        let u32_val = (((self.buffer[self.offset] as u32) << 24) |
                       ((self.buffer[self.offset + 1] as u32) << 16) |
                       ((self.buffer[self.offset + 2] as u32) << 8) |
                       ((self.buffer[self.offset + 3] as u32)));
        self.offset += 4;

        u32_val
    }
    pub fn decode_macaddr(&mut self, mac_ptr : &mut [u8; 6]) {
        mac_ptr[0] = self.buffer[self.offset];
        mac_ptr[1] = self.buffer[self.offset] + 1;
        mac_ptr[2] = self.buffer[self.offset] + 2;
        mac_ptr[3] = self.buffer[self.offset] + 3;
        mac_ptr[4] = self.buffer[self.offset] + 4;
        mac_ptr[5] = self.buffer[self.offset] + 5;

        self.offset += 6;
    }

    pub fn set_len(&mut self, len : usize) { self.rx_len = len; }
}

File: common/pkt_buffer.c
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <pthread.h>

#include "pkt_buffer.h"
#include "event_info.h"

void pkt_buffer_initialize(pkt_buffer_t *pkt_buf)
{
    pkt_buf->offset     = 0;
    pkt_buf->rx_len     = 0;
    pkt_buf->tx_len     = 0;
    pkt_buf->ref_count  = 0;
    pthread_mutex_init(&pkt_buf->lock, NULL);
    NETOS_EVENT_INIT(pkt_buf->event_type, pkt_buf->event_desc);
}

void pkt_buffer_ref_count_up(pkt_buffer_t *pkt_buf)
{
    pthread_mutex_lock(&pkt_buf->lock);
    pkt_buf->ref_count ++;
    pthread_mutex_unlock(&pkt_buf->lock);
}

void pkt_buffer_ref_count_up_n(pkt_buffer_t *pkt_buf, uint32_t count)
{
    pthread_mutex_lock(&pkt_buf->lock);
    pkt_buf->ref_count += 2;
    pthread_mutex_unlock(&pkt_buf->lock);
}

void pkt_buffer_ref_count_down(pkt_buffer_t *pkt_buf)
{
    pthread_mutex_lock(&pkt_buf->lock);
    if (pkt_buf->ref_count > 0) {
        pkt_buf->ref_count --;
    }
    pthread_mutex_unlock(&pkt_buf->lock);
}

bool pkt_buffer_rx_frame_in_range(pkt_buffer_t *pkt_buf, uint32_t hdr_len)
{
    return (pkt_buf->rx_len >= (pkt_buf->offset + hdr_len));
}

void pkt_buffer_decode_byte(pkt_buffer_t *pkt_buf, uint8_t *u8)
{
    *u8 = pkt_buf->buffer[pkt_buf->offset];
    pkt_buf->offset ++;
}

void pkt_buffer_encode_byte(pkt_buffer_t *pkt_buf, uint8_t u8)
{
    pkt_buf->buffer[pkt_buf->offset] = u8;
    pkt_buf->offset ++;
}

void pkt_buffer_decode_2_bytes(pkt_buffer_t *pkt_buf, uint16_t *u16)
{
    uint8_t *buf = pkt_buf->buffer;
    uint32_t offset = pkt_buf->offset;

    *u16 = (buf[offset] << 8) | (buf[offset + 1]);
    pkt_buf->offset += 2;
}

void pkt_buffer_encode_2_bytes(pkt_buffer_t *pkt_buf, uint16_t u16)
{
    pkt_buf->buffer[pkt_buf->offset] = (u16 & 0xFF00) >> 8;
    pkt_buf->buffer[pkt_buf->offset + 1] = (u16 & 0x00FF);
    pkt_buf->offset += 2;
}

void pkt_buffer_decode_4_bytes(pkt_buffer_t *pkt_buf, uint32_t *u32)
{
    uint8_t *buf = pkt_buf->buffer;
    uint32_t offset = pkt_buf->offset;

    *u32 = (buf[offset] << 24) | (buf[offset + 1] << 16) |
           (buf[offset + 2] << 8) | (buf[offset + 3]);
    pkt_buf->offset += 4u;
}

void pkt_buffer_encode_4_bytes(pkt_buffer_t *pkt_buf, uint32_t u32)
{
    pkt_buf->buffer[pkt_buf->offset] = (u32 & 0xFF000000) >> 24;
    pkt_buf->buffer[pkt_buf->offset + 1] = (u32 & 0x00FF0000) >> 16;
    pkt_buf->buffer[pkt_buf->offset + 2] = (u32 & 0x0000FF00) >> 8;
    pkt_buf->buffer[pkt_buf->offset + 3] = (u32 & 0x000000FF);
    pkt_buf->offset += 4;
}

void pkt_buffer_decode_bytes(pkt_buffer_t *pkt_buf, uint8_t *data, uint32_t data_len)
{
    memcpy(data, &pkt_buf->buffer[pkt_buf->offset], data_len);
    pkt_buf->offset += data_len;
}

void pkt_buffer_encode_bytes(pkt_buffer_t *pkt_buf, uint8_t *data, uint32_t data_len)
{
    memcpy(&pkt_buf->buffer[pkt_buf->offset], data, data_len);
    pkt_buf->offset += data_len;
}

